scripts/run_integration.sh

```diff
@@ -0,0 +1,173 @@
+#!/bin/bash
+
+readonly SCRIPT_FOLDER=$(cd -P -- "$(dirname -- "$0")" && pwd -P)
+readonly PROJECT_FOLDER="${SCRIPT_FOLDER}/.."
+readonly ARCHIVE_FOLDER=~/archives
+readonly QEMU_INSTALL=${ARCHIVE_FOLDER}/qemu
+readonly DEFAULT_CMAKE_ARGS=" -DCMAKE_BUILD_TYPE=Debug -DBUILD_TESTING=ON"
+
+function extract() {
+  case $1 in
+    *.tar.bz2)   tar xjf $1    ;;
+    *.tar.xz)    tar xJf $1    ;;
+    *.tar.gz)    tar xzf $1    ;;
+    *)
+      echo "don't know how to extract '$1'..."
+      exit 1
+  esac
+}
+
+function unpackifnotexists() {
+  mkdir -p ${ARCHIVE_FOLDER}
+  cd ${ARCHIVE_FOLDER}
+  local URL=$1
+  local DESTINATION=`pwd`/$2
+  if [[  ! -d "${DESTINATION}" ]] ; then
+    local ARCHIVE_NAME=`echo ${URL} | sed 's/.*\///'`
+    test -f ${ARCHIVE_NAME} || wget ${URL}
+    extract ${ARCHIVE_NAME}
+  fi
+}
+
+function installqemuifneeded() {
+  local VERSION=${QEMU_VERSION:=2.11.1}
+  local ARCHES=${QEMU_ARCHES:=arm aarch64 i386 x86_64 mips mipsel}
+  local TARGETS=${QEMU_TARGETS:=$(echo $ARCHES | sed 's#$# #;s#\([^ ]*\) #\1-linux-user #g')}
+
+  if echo "${VERSION} ${TARGETS}" | cmp --silent ${QEMU_INSTALL}/.build -; then
+    echo "qemu ${VERSION} up to date!"
+    return 0
+  fi
+
+  echo "VERSION: ${VERSION}"
+  echo "TARGETS: ${TARGETS}"
+
+  rm -rf ${QEMU_INSTALL}
+
+  # Checking for a tarball before downloading makes testing easier :-)
+  local QEMU_URL="http://wiki.qemu-project.org/download/qemu-${VERSION}.tar.xz"
+  local QEMU_FOLDER="qemu-${VERSION}"
+  unpackifnotexists ${QEMU_URL} ${QEMU_FOLDER}
+  cd ${QEMU_FOLDER}
+
+  ./configure \
+    --prefix="${QEMU_INSTALL}" \
+    --target-list="${TARGETS}" \
+    --disable-docs \
+    --disable-sdl \
+    --disable-gtk \
+    --disable-gnutls \
+    --disable-gcrypt \
+    --disable-nettle \
+    --disable-curses \
+    --static
+
+  make -j4
+  make install
+
+  echo "$VERSION $TARGETS" > ${QEMU_INSTALL}/.build
+}
+
+function assert_defined(){
+  local VALUE=${1}
+  : ${VALUE?"${1} needs to be defined"}
+}
+
+function integrate() {
+  cd ${PROJECT_FOLDER}
+  cmake -H. -B${BUILD_DIR} ${DEFAULT_CMAKE_ARGS} ${CMAKE_ADDITIONAL_ARGS}
+  cmake --build ${BUILD_DIR} --target all
+
+  if [[ -n "${QEMU_ARCH}" ]]; then
+    if [[ "${QEMU_ARCH}" == "DISABLED" ]]; then
+      QEMU="true || "
+    else
+      installqemuifneeded
+      QEMU="${QEMU_INSTALL}/bin/qemu-${QEMU_ARCH} ${QEMU_ARGS}"
+    fi
+  else
+    QEMU=""
+  fi
+  # Run tests
+  for test_binary in ${BUILD_DIR}/test/*_test; do ${QEMU} ${test_binary}; done
+  # Run demo program
+  ${QEMU} ${BUILD_DIR}/list_cpu_features
```

Comment:
I would run the demo first to get the qemu config, then run the test knowing the qemu "environment"...

Reply:
I'm not sure I understand what you mean. They both use QEMU (which may be empty).